qt-build/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Be Wilson <be.wilson@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! This crate provides information about the Qt installation and can invoke Qt's
+//! [moc](https://doc.qt.io/qt-6/moc.html) code generator. This crate does not build
+//! any C++ code on its own. It is intended to be used in [build.rs scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html)
+//! together with
+//! [cc](https://docs.rs/cc/latest/cc/),
+//! [cxx_build](https://docs.rs/cxx-build/latest/cxx_build/), or
+//! [cpp_build](https://docs.rs/cpp_build/latest/cpp_build/).
+
+use std::{
+    env,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+pub use versions::SemVer;
+
+/// Helper for build.rs scripts using Qt
+/// ```
+/// let qt_modules = vec!["Core", "Gui"]
+///     .iter()
+///     .map(|m| String::from(*m))
+///     .collect();
+/// let qtbuild = qt_build::QtBuild::new(qt_modules).expect("Could not find Qt installation");
+/// ```
+pub struct QtBuild {
+    pub version: SemVer,
+    qmake_executable: String,
+    qt_modules: Vec<String>,
+}
+
+impl QtBuild {
+    /// Search for where Qt is installed using qmake. Specify the Qt modules you are
+    /// linking with the `qt_modules` parameter, ommitting the `Qt` prefix (`"Core"`
+    /// rather than `"QtCore"`). After construction, use the [QtBuild::qmake_query]
+    /// method to get information about the Qt installation.
+    ///
+    /// The directories specified by the `PATH` environment variable are where qmake is
+    /// searched for. Alternatively, the `QMAKE` environment variable may be set to specify
+    /// an explicit path to qmake.
+    ///
+    /// If multiple major versions (for example, `5` and `6`) of Qt could be installed, set
+    /// the `QT_VERSION_MAJOR` environment variable to force which one to use. When using Cargo
+    /// as the build system for the whole build, prefer using `QT_VERSION_MAJOR` over the `QMAKE`
+    /// environment variable because it will account for different names for the qmake executable
+    /// that some Linux distributions use.
+    ///
+    /// However, when building a Rust staticlib that gets linked to C++ code by a C++ build
+    /// system, it is best to use the `QMAKE` environment variable to ensure that the Rust
+    /// staticlib is linked to the same installation of Qt that the C++ build system has
+    /// detected. With CMake, you can get this from the `Qt${QT_VERSION_MAJOR}::qmake`
+    /// target's `IMPORTED_LOCATION` property, for example:
+    /// ```
```

Comment:
could put ```cmake as formatters might have colouring for cmake syntax ?

Reply:
rustdoc doesn't have syntax highlighting for CMake, but I suppose it doesn't hurt to add that anyway